### PR TITLE
Add new database_secrets_mount resource

### PR DESCRIPTION
**Pull request by benashz**
This PR introduces a new resource for provisioning multiple database secret engines under a dedicated database mount. It is the combination of `vault_mount` and `vault_database_secret_backend_connection`. 

#### Caveats:
The `vault_database_secrets_mount` resource will be replaced

- On update, a database engine block is removed
- On update, the `name` is changed for any configured engine

#### Example usage:
```hcl
resource "vault_database_secrets_mount" "db" {
  path = "db"

  mssql {
    name           = "db1"
    username       = "sa"
    password       = "super_secret_1"
    connection_url = "sqlserver://{{username}}:{{password}}@127.0.0.1:1433"
    allowed_roles = [
      "dev1",
    ]
  }

  postgresql {
    name              = "db2"
    username          = "postgres"
    password          = "super_secret_2"
    connection_url    = "postgresql://{{username}}:{{password}}@127.0.0.1:5432/postgres"
    verify_connection = true
    allowed_roles = [
      "dev2",
    ]
  }
}

resource "vault_database_secret_backend_role" "dev1" {
  name    = "dev1"
  backend = vault_database_secrets_mount.db.path
  db_name = vault_database_secrets_mount.db.mssql[0].name
  creation_statements = [
    "CREATE LOGIN [{{name}}] WITH PASSWORD = '{{password}}';",
    "CREATE USER [{{name}}] FOR LOGIN [{{name}}];",
    "GRANT SELECT ON SCHEMA::dbo TO [{{name}}];",
  ]
}

resource "vault_database_secret_backend_role" "dev2" {
  name    = "dev2"
  backend = vault_database_secrets_mount.db.path
  db_name = vault_database_secrets_mount.db.postgresql[0].name
  creation_statements = [
    "CREATE ROLE \"{{name}}\" WITH LOGIN PASSWORD '{{password}}' VALID UNTIL '{{expiration}}';",
    "GRANT SELECT ON ALL TABLES IN SCHEMA public TO \"{{name}}\";",
  ]
}
```

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note

```

Output from acceptance testing:

```
$ make testacc TESTARGS='-v -test.run TestAccDatabaseSecretsMount_mssql*'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test -v -v -test.run TestAccDatabaseSecretsMount_mssql* -timeout 30m ./...

=== RUN   TestAccDatabaseSecretsMount_mssql
--- PASS: TestAccDatabaseSecretsMount_mssql (10.47s)
=== RUN   TestAccDatabaseSecretsMount_mssql_multi
--- PASS: TestAccDatabaseSecretsMount_mssql_multi (17.30s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     34.174s
```

### CODE REVIEW

File: vault/resource_database_secrets_mount.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+		Description: "Total number of database secret engines configured under the mount.",
+	}
+
+	return s
+}
+
+func addCommonDatabaseSchema(s *schema.Schema) {
+	elem := s.Elem.(*schema.Resource)
+	for k, v := range getCommonDatabaseSchema() {
+		// TODO handle intersection errors
```
Comment (vinay-gopalan) [inline vault/resource_database_secrets_mount.go:137]: Is this still a concern?

File: vault/resource_database_secrets_mount.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+func databaseSecretsMountRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	if err := readMount(d, meta, true); err != nil {
+		return err
+	}
+
+	root := d.Id()
+	if root == "" {
```
Comment (vinay-gopalan) [inline vault/resource_database_secrets_mount.go:266]: Wondering why this is not an error. If we're in the Read func, should there be a case where the ID on the resource isn't set?

File: vault/resource_database_secrets_mount.go [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+func readDBEngineConfig(d *schema.ResourceData, client *api.Client, store *dbConfigStore, name string) error {
+	root := d.Id()
+
+	path := databaseSecretBackendConnectionPath(root, name)
+	log.Printf("[DEBUG] Reading database connection config %q", path)
+	resp, err := client.Logical().Read(path)
+	if err != nil {
+		return fmt.Errorf("error reading database connection config %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Read database connection config %q", path)
```
Comment (vinay-gopalan) [inline vault/resource_database_secrets_mount.go:310]: nit: per a previous conversation of ours, could we make this `Successfully read...` for better readability and consistency with other debug strings?

File: vault/resource_mount.go [hunk 4]
```diff
@@ -10,7 +10,112 @@ import (
...
+		},
+	}
+	for _, v := range excludes {
+		delete(s, v)
+	}
+	return s
+}
+
 func MountResource() *schema.Resource {
+	// TODO: refactor to shared schema shared with resource_database_mount
```
Comment (vinay-gopalan) [inline vault/resource_mount.go:118]: Is this still a TODO?

File: website/docs/r/database_secrets_mount.md [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+
+* `username` - (Optional) The username to be used in the connection (the account admin level).
+
+* `password` - (Optional) The password to be used in the connection.
+
+* `username_template` - (Optional) - [Template](https://www.vaultproject.io/docs/concepts/username-templating) describing how dynamic usernames are generated.
+
+## Attributes Reference
+
+No additional attributes are exported by this resource.
```
Comment (vinay-gopalan) [inline website/docs/r/database_secrets_mount.md:440]:
```suggestion

```

File: website/docs/r/database_secrets_mount.md [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+sidebar_current: "docs-vault-resource-database-secrets-mount"
+description: |-
+  Configures any number of database secrets engines under a single mount resource
+---
+
+# vault\_database\_secrets\_mount
+
+Configure any number of database secrets engines under a single dedicated mount resource.
+
+~> **Important** All data provided in the resource configuration will be
```
Comment (vinay-gopalan) [inline website/docs/r/database_secrets_mount.md:13]: nit: I've only seen this paragraph on Datasource docs and haven't seen it on Resource docs

File: vault/resource_database_secret_backend_connection.go [hunk 7]
```diff
@@ -148,6 +148,10 @@ func (i *dbEngine) Name() string {
 	return i.name
```
Comment (vinay-gopalan) [inline vault/resource_database_secret_backend_connection.go:148]: These aren't part of your changes, but I'm wondering if we need both `dbEngine.Name()` and `dbEngine.String()`. Looks like they are duplicate functions

**Review by vinay-gopalan — COMMENTED** [review]
Looking great! A few minor suggestions/queries

File: vault/resource_database_secrets_mount.go [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+		return nil
+	}
+
+	return s.result[db.Name()]
+}
+
+func (s *dbConfigStore) Result() map[string][]map[string]interface{} {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	if s.result == nil {
```
[the remark this answers is not in the document]
Comment (benashz) [reply vault/resource_database_secrets_mount.go:46]: good call, I will have it return a copy as well.

File: vault/resource_database_secrets_mount.go (hunk 1 above)
Comment (benashz) [reply vault/resource_database_secrets_mount.go:137]: Not a huge concern IMO, I think we can punt on it for now.

File: vault/resource_database_secrets_mount.go (hunk 2 above)
Comment (benashz) [reply vault/resource_database_secrets_mount.go:266]: Yeah, sort of misleading. Essentially `readMount()` takes care of setting to `""` if the mount was not found.

Comment (benashz) [reply vault/resource_database_secrets_mount.go:266]: Added a comment.

File: vault/resource_mount.go (hunk 4 above)
Comment (benashz) [reply vault/resource_mount.go:118]: good catch, it's obsolete.

File: website/docs/r/database_secrets_mount.md (hunk 5 above)
Comment (benashz) [reply website/docs/r/database_secrets_mount.md:440]: Good catch!

File: website/docs/r/database_secrets_mount.md (hunk 6 above)
Comment (benashz) [reply website/docs/r/database_secrets_mount.md:13]: interesting, I based the docs of off `database_secret_backend_connection.md`. We should probably revisit at a later date.

File: vault/resource_database_secret_backend_connection.go (hunk 7 above)
Comment (benashz) [reply vault/resource_database_secret_backend_connection.go:148]: Yeah, `String()` should probably return `Name()`. `String()` implements the `Stringer` interface

File: vault/resource_database_secrets_mount.go (hunk 2 above)
Comment (vinay-gopalan) [reply vault/resource_database_secrets_mount.go:271]: Thanks!

**Review by vinay-gopalan — APPROVED** [review]
Looks good!

**Pull request merged**